### ISSUE TITLE
Help Needed: Can't get past 'No source found' error

### ISSUE THREAD

**Issue by ernieturner** [monounity/karma-typescript#79]
karma-typescript version: `2.1.7`

I think this might be somewhat similar to #75, but I can't seem to figure out what I have configured incorrectly that I'm getting this error:

```
02 02 2017 11:33:20.684:INFO [compiler.karma-typescript]: Compiling project using Typescript 2.1.5
02 02 2017 11:33:22.706:INFO [compiler.karma-typescript]: Compiled 1 files in 1475 ms.
02 02 2017 11:33:22.712:ERROR [karma]: Error: No source found for [...]/app/components/atom/tests/Button.test.tsx!
Is there a mismatch between the Typescript compiler options and the Karma config?
```

My karma config file looks like this:

```
module.exports = function(config) {
    config.set({
        basePath: '.',
        frameworks: ["jasmine", "karma-typescript"],
        files: [
            { pattern: "app/**/*.+(ts|tsx)" }
        ],
        preprocessors: {
            "**/*.+(ts|tsx)": ["karma-typescript"]
        },
        reporters: ["progress", "karma-typescript"],
        browsers: ["PhantomJS"],
        singleRun: true,
        karmaTypescriptConfig: {
            tsconfig: './tsconfig.json',
            coverageOptions: {
                exclude: /\.test\.tsx?/
            },
            compilerOptions: {
                jsx: "react"
            }
        }
    });
};
```

And my tsconfig.json looks like this:

```
{
    "compilerOptions": {
        "sourceMap": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strictNullChecks": true,
        "noUnusedParameters": true,
        "noUnusedLocals": true,
        "module": "commonjs",
        "target": "es6",
        "jsx": "preserve",
        "baseUrl": ".",
        "paths": {
            "*": [
                "app/*",
                "app/components/*"
            ]
        },
        "moduleResolution": "node"
    },
    "include": [
        "./app/*"
    ],
    "exclude": [
        "node_modules"
    ],
    "awesomeTypescriptLoaderOptions": {
        "useCache": true
    }
}
```

I only have a single file that I'm trying to get setup to test thats under `app/components/atom/Button.tsx` and the test file is at `app/components/atom/tests/Button.tsx`. 

Please let me know if theres any other info that would help. Thanks in advance!

**Comment by erikbarke** [comment]
Hey @ernieturner, if you turn on debug logging in your Karma conf, `logLevel: config.LOG_DEBUG` the compiler will tell you which files it has picked up (looks like it only picked up one file), is "Button.test.tsx" listed there?

**Comment by ernieturner** [comment]
Hi @erikbarke, thanks for taking a look. Here's a bit of the output that I'm getting:

`
[preprocessor.karma-typescript]: Processing ".../app/actions/AppGroupActions.ts".
[preprocessor.karma-typescript]: Processing ".../app/actions/ModalActions.ts".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/Button.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/AppGroups/AppGroupList.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/ErrorMessage.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/LoadingPlaceholder.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/Logo.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/Modal.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/TagChip.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/atom/tests/Button.test.tsx".
[preprocessor.karma-typescript]: Processing ".../app/components/containers/ModalWrapper.tsx".
....
....
[compiler.karma-typescript]: Reading: .../app/main.tsx
[compiler.karma-typescript]: Reading: .../node_modules/@types/react/index.d.ts
[compiler.karma-typescript]: Reading: .../node_modules/@types/react-dom/index.d.ts
[compiler.karma-typescript]: Reading: .../node_modules/redux/index.d.ts
....
...
[compiler.karma-typescript]: Compiled 1 files in 1832 ms.
02 02 2017 16:24:46.979:ERROR [karma]: Error: No source found for .../app/components/atom/tests/Button.test.tsx!
`

So it looks like it's at least precompiling all of the files in my tree and includes both the Button.tsx and Button.test.tsx in there as well.

**Comment by erikbarke** [comment]
Karma is picking up a lot of files here, but the compiler only compiles one file, `Compiled 1 files in 1832 ms.`, so I'm thinking this must a `tsconfig` issue, does it work if you modify the `include` section to include subfolders?

```
"include": [
    "app/**/*"
],
```

**Comment by ernieturner** [comment]
Yep, that definitely fixed the problem. After trying a few other combinations of configurations I figured out that I don't actually need that line in my `tsconfig`. Thanks a bunch for the help on this, it's really appreciated.

Just in case it helps others, here's what my final configs look like:

`tsconfig`
```
{
    "compilerOptions": {
        "sourceMap": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strictNullChecks": true,
        "noUnusedParameters": true,
        "noUnusedLocals": true,
        "module": "commonjs",
        "target": "es6",
        "jsx": "preserve",
        "baseUrl": ".",
        "paths": { //Allow for relative file path includes from the following as roots. Lets me include as if I'm relative to these dirs.
            "*": [
                "app/*",
                "app/components/*"
            ]
        },
        "moduleResolution": "node"
    },
    "exclude": [
        "node_modules"
    ],
    "awesomeTypescriptLoaderOptions": {
        "useCache": true
    }
}
```

`karma.conf.js`
```

**Comment by adamcogx** [comment]
I'm seeing a form of this error.  If karma is running in watch mode (singleRun: false) and I add a new .spec.ts file in Visual Studio, it will crash Karma with this error as the file is empty when first created.  After I restart Karma, it picks up the test and works as expected.

I'm running with the latest from master (as per your request in #93 )

**Comment by adamcogx** [comment]
It's not a matter of it being empty.  I think it has to do with a new file being added to the set of files to compile.  It compiles it fine, but because its not a registered entry point (which seems to be set when karma starts) it can't find it when it goes to get it.  I've tried following the advice on this page (add an include: []) but that doesn't solve it.

**Comment by erikbarke** [comment]
On my laptop new files get picked up by Karma in watch mode, but the compiler doesn't realize there's a new file, hence the `No source found for...`. Karma has a file but the compiler doesn't have a matching `sourceFile`, this is a bug.

**Comment by erikbarke** [comment]
@adamcogx, I've added support for file system changes, does it work for you if you install from the master branch, `npm i github:monounity/karma-typescript`?

**Comment by adamcogx** [comment]
If you add a file, everything is fine (it re-evaluates and loads the file).  One thing that happens is if you delete a file, you get:

```
Chrome 57.0.2987 (Windows 10 0.0.0) ERROR
Uncaught Error: Can't find entrypoint [../app/tests/fuzzy.ts] (required by commonjs.js)  at node_modules/karma-typescript/src/client/commonjs.js:13
```

it doesn't crash the system (which it used to do) but it fails to re-run the tests after the delete (it loads all the files as I can see console logs of my angular stuff booting up, but fails to run the actual .spec.ts files).  It requires changing an existing file to get it to run the tests again.

**Comment by erikbarke** [comment]
I can't reproduce this error, I've tried on OSX and Windows 10 with the latest from master. Could you enable debug logging in Karma and post the output here please?

**Comment by zaktimson** [comment]
I am having a simillar issue. I have been trying to get unit tests going with my typescript library
```
08 05 2017 15:16:16.876:INFO [compiler.karma-typescript]: Compiled 4 files in 1730 ms.
08 05 2017 15:16:16.877:ERROR [karma]: Error: No source found for C:/Users/zakti/Documents/LocalStorage/src/decorators.spec.d.ts!
Is there a mismatch between the Typescript compiler options and the Karma config?
```

Folder Layout:
```
- node_modules
- src
    |- decorators.spec.ts
    |- decorators.ts
    | - storage.ts
- index.ts
- karma.conf.js
- package.json
- tsconfig.json
- typings.json
```

karma.conf.js:
```
module.exports = function (config) {
    config.set({
        basePath: '',
        frameworks: ['jasmine', 'karma-typescript'],
        files: [
            {pattern: 'src/**/*.ts'}
        ],
        exclude: [],
        preprocessors: {
            '**/*.ts': ['karma-typescript']
        },
        reporters: ['progress'],
        port: 9876,
        colors: true,
        logLevel: config.LOG_DEBUG,
        autoWatch: false,
        browsers: ['PhantomJS'],
        singleRun: false,
        concurrency: Infinity
    })
};

```

tsconfig.json
```
{
  "compilerOptions": {
    "declaration": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitAny": false,
    "removeComments": false,
    "sourceMap": true,
    "target": "es5"
  }
}
```

**Comment by erikbarke** [comment]
Hey @zaktimson, could you run your project with `logLevel: config.LOG_DEBUG` and attach the output as a text file here please? I'd like to know why the compiler is looking for the file `decorators.spec.d.ts`.

**Comment by zaktimson** [comment]
@erikbarke It seems to change, I have had it complain about the definitions file for all 4 ts files.
[debug_log.txt](https://github.com/monounity/karma-typescript/files/987671/debug_log.txt)

**Comment by erikbarke** [comment]
@zaktimson, Karma does pick up your `.d.ts` files but the Typescript compiler doesn't (which it does on my laptop...). Does it work if you exclude all `.d.ts` files in the Karma config?

```js
exclude: ["src/**/*.d.ts"]
```

**Comment by zaktimson** [comment]
Yes that seemed to fix it! Thank you very much. I have been trying for a week to get my unit tests going.
